### PR TITLE
[libc++] LWG2381: Inconsistency in parsing floating point numbers

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR implements [LWG2381](https://cplusplus.github.io/LWG/issue2381) by rejecting `'i'`, `'I'`, `'n'`, `'N'` in FP parsing, as inf and NaN are intendedly rejected by that LWG issue.

The source character array used for parsing is `"0123456789abcdefABCDEFxX+-pPiInN"`, whose first 26 or 28 characters
are used for parsing integers or floating-point values respectively. Previously, libc++ used 32 characters, including `'i'`, `'I'`, `'n'`, `'N'`, for FP parsing, which was inconsistent with LWG2381. This PR also replaces magic numbers 26 and 28 (formerly 32) with named constants.

Drive-by change: when the first character (possibly after the leading `'+'` or `'-'`) is not a decimal digit but an acceptable character (e.g., `'p'` or `'e'`), the character is not accumulated now (per Stage 2 in [facet.num.get.virtuals]/3).

#65168 may be rendered invalid, see https://github.com/llvm/llvm-project/pull/65168#issuecomment-1868533342.

Apple back-deployment targets remain broken, likely due to dylib. XFAIL is marked in related tests.

### CODE REVIEW

File: libcxx/include/locale
Comment (mordante) [inline libcxx/include/locale:550]: Why is this changed from 32 to 28. Please add the reason to the commit message too.

File: libcxx/include/locale
Comment (mordante) [inline libcxx/include/locale:370]: Thanks for removing magic numbers, can you add some comment what these values are use for?

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp
Comment (mordante) [inline libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp:173]: `1.189731495357231765021264e+49321` does not fit in binary128 too.
What changed that these values no longer work on MSVC?

**Review by mordante — COMMENTED** [review]
Thanks for working on this. I've a few questions before I want to look deeper in the changes.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp:173]: Thanks for the correction. I've verified that `1.189731495357231765021264e+49321` and `1.189731495357231765021264e+49329` consistently result in infinity on MSVC, so perhaps these cases shouldn't be guarded.

File: libcxx/include/locale
Comment (frederick-vs-ja) [reply libcxx/include/locale]: Added.

**Comment by frederick-vs-ja** [comment]
@mordante ping. It seems that this missed 18.0.

**Comment by mordante** [comment]
> @mordante ping. It seems that this missed for 18.0.

Thanks for the reminder, it slipped under radar. I'll try to have a look soon.

File: libcxx/include/locale
Comment (mordante) [inline libcxx/include/locale:370]: I would move these down, directly below `static const char __src[33]; // "0123456789abcdefABCDEFxX+-pPiInN"`
```suggestion
  static const size_t __int_chr_cnt = 26; // count of leading characters in __src used for parsing integers ("012..X+-")
  static const size_t __fp_chr_cnt  = 28; // count of leading characters in __src used for parsing floating-point values  ("012..-pP")
```
I think adding a range makes it easier to understand which characters are in the range.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp
Comment (mordante) [inline libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp:13]: This is the new way to do this. I expect you need to rebase this patch to be able to use this.
```suggestion
// UNSUPPORTED: using-built-library-before-llvm-19
```

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp
Comment (mordante) [inline libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp:188]: Can you also test with "p00", "+p00", "-p00" and similar for "P00", "e00", and "E00"?

**Review by mordante — COMMENTED** [review]
Thanks for the patch! In general this looks good. Could you apply the changes to the `double` test to `float` and `long double` too? After these changes are applied things look good to me.

File: libcxx/include/locale
Comment (frederick-vs-ja) [reply libcxx/include/locale]: Done. (Although the current formatting for `__fp_chr_cnt` is a bit ugly due to the long comment...)

Comment (mordante) [reply libcxx/include/locale]: I agree, maybe move the comment before the declaration instead of a trailing comment.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_float.pass.cpp
Comment (mordante) [inline libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_float.pass.cpp:185]: For consistency with the existing code.
```suggestion
      assert(v == 0.0f);
```

Please make a similar changes for the new `long double` tests.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp
Comment (mordante) [inline libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp:322]: Mainly curious why is this change needed after this patch?

**Review by mordante — COMMENTED** [review]
In general LGTM with some minor nits. I've posted one question.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp:187]: Oh, these stuffs are now completely accumulated, which seems to be pre-existing. Presumably the leading 'p' or 'e' shouldn't be accumulated ([[facet.num.get.virtuals]/3](https://eel.is/c++draft/facet.num.get.virtuals#3.2.sentence-5)).

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp:322]: This is a drive-by change that would make MSVC, whose `long double` is of binary64 format, pass the test.

Comment (mordante) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp:322]: I initially expected that, but these tests are running with clang-cl (and mingw). I expect at least clang-cl uses LDBL_MANT_DIG == 53 and thus give the same results as MSVC STL.

I'm especially puzzled why these infinity tests fail while the infinity test above passes.
The `304888344611713860501504000000` test seems more likely to fail due to a lack of precision.
(I see this test is listed in https://github.com/microsoft/STL/blob/main/tests/libcxx/expected_results.txt)

Did you run these tests with MSVC STL and checked why they failed?

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp:187]: I'll try to fix the checking for the leading sign and digit. But there'rs still non-trivial issues, which perhaps should be fixed in another PR. Reported #87969.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_long_double.pass.cpp]: Thanks for noticing this! I've confirmed that the conditional compilation has become unnecessary now.

File: libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp
Comment (frederick-vs-ja) [reply libcxx/test/std/localization/locale.categories/category.numeric/locale.num.get/facet.num.get.members/get_double.pass.cpp:187]: > Can you also test with "p00", "+p00", "-p00" and similar for "P00", "e00", and "E00"?

These cases are now fixed.

**Review by mordante — APPROVED** [review]
LGTM!

**Comment by bgra8** [comment]
@mordante this patch causes strings like ".5" to no longer parse and this is most likely a bug.

Here's a small reproducer:

```
#include <sstream>
#include <cassert>

int main(int , char** ) {
  std::istringstream strm(".5");
  float val = 0;
  strm >> val;

  assert(!strm.fail());
  assert(strm.eof());
  assert(val == .5);
  return 0;
}
```

Can you please take a look?

**Comment by frederick-vs-ja** [comment]
It seems that I made a wrong change. Decimal point `'.'` should also be allowed to be the leading character.

https://github.com/llvm/llvm-project/blob/f34dedbf44a0578430f81de94f294f4474fee0fb/libcxx/include/locale#L986-L998

I'll fix the regression later today.

**Comment by bgra8** [comment]
@frederick-vs-ja can you please revert first to unblock us? You can then work without pressure to fix the issue and avoid introducing other new issues.

@mordante sorry for misattributing this to you!

**Comment by frederick-vs-ja** [comment]
I'm not a big fan of revert-and-reland. Given both reverting and a direct fix require review, I'd like to open a PR for direct fix.

If maintainers prefers reverting, I'll change the PR to revert.

**Pull request merged**
